cluster/core/src/main/scala/com/lightbend/lagom/internal/cluster/JoinClusterImpl.scala

```diff
@@ -74,20 +75,39 @@ private[lagom] object JoinClusterImpl {
-          .shutdownReason().contains(CoordinatedShutdown.ClusterDowningReason)
+        val shutdownReason: Option[Reason] = CoordinatedShutdown(system).shutdownReason()
 
-        // if 'exitJvm' is enabled and the trigger (aka Reason) for CoordinatedShutdown is ClusterDowning
-        // we must exit the JVM. This can lead to the cluster closing before the `Application` but we're
-        // out of the cluster (downed) already so the impact is acceptable. This will be improved when
-        // Play delegates all shutdown logic to CoordinatedShutdown in Play 2.7.x.
+        val reasons: Seq[Reason] = Seq(
+          ClusterDowningReason,
+          ClusterJoinUnsuccessfulReason,
+          IncompatibleConfigurationDetectedReason)
+
+        val reasonIsDowning: Boolean = shutdownReason.exists(reasons.contains)
+
+        // if 'exitJvm' is enabled and the trigger (aka Reason) for CoordinatedShutdown is ClusterDowning,
+        // JoinUnsuccessful, etc... we must exit the JVM. This can lead to the cluster closing before
+        // the `Application` but we're out of the cluster (downed) already so the impact is acceptable.
         if (exitJvm && reasonIsDowning) {
-          // In case ActorSystem shutdown takes longer than 10 seconds,
-          // exit the JVM forcefully anyway.
-          // We must spawn a separate thread to not block current thread,
-          // since that would have blocked the shutdown of the ActorSystem.
+
+          // If this code is running, it means CoordinatedShutdown was triggered. CoordinatedShutdown of
+          // a given actor system can only be invoked once: further invocations block until the initial
+          // one completes. The code below works as following:
+          //   - create a new Thread and invoke System.exit(-1)
+          //   - terminate the "exit-jvm-when-downed" task
+          //   - in parallel, the invocation of System.exit(-1) has triggered the execution of the JVM shutdown Hooks
+          //   - Play's JVM shutdown hooks proceed to stop Play which means stopping the Server and the
+          //     Application. That also involves Play's ApplicationLifecycle. In a particular step of that stop
+          //     process Play must stop its Actor System so it invokes CoordinatedShutdown. Since the
+          //     CoordinatedShutdown is already running, that operation blocks.
+          // !! At this point, there's a CoordinatedShutdown running (triggered by a Downing event) and a
+          //    reference to a CoordinatedShutdown blocked until the run completes. The thread blocked is the
+          //    JVM shutdown thread.
+          //  - When the main CoordinatedShutdown completes, the JVM keeps running because Play (and Lagom)
+          //    tune Akka's `exit-jvm=off`.
+          //  - When the main CoordinatedShutdown completes, the thread blocked on the main CoordinatedShutdown
+          //    unblocks and completes the execution of `System.exit`.
```

Comment:
This section of the changes is a complete review/update/clarification on how/why the shutdown works in case of a cluster-related reason.